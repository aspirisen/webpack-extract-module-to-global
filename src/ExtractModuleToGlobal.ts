import webpack from "webpack";

export class ExtractModuleToGlobal {
  public static pluginName = "ExtractModuleToGlobal";

  public static loader = require.resolve("./ExtractModuleToGlobalLoader");

  private trackPaths: Record<string, [content: string, sourceMap: any]> = {};

  public addSource(path: string, content: string, map: any) {
    this.trackPaths[path] = [content, map];
  }

  public apply(compiler: webpack.Compiler) {
    compiler.hooks.thisCompilation.tap(
      ExtractModuleToGlobal.pluginName,
      (compilation) => {
        const jsHooks =
          webpack.javascript.JavascriptModulesPlugin.getCompilationHooks(
            compilation
          );

        jsHooks.render.tap(
          ExtractModuleToGlobal.pluginName,
          (source, renderContext) => {
            if (!(source instanceof webpack.sources.ConcatSource)) {
              return source;
            }

            const finalSource = new webpack.sources.ConcatSource();
            const modules = Array.from(
              renderContext.chunkGraph.getChunkModulesIterable(
                renderContext.chunk
              )
            );

            modules.forEach((m) => {
              const resourceName = (m as any).resource;
              const resource = this.trackPaths[resourceName];

              if (!resource) {
                return;
              }

              const [content, sourceMap] = resource;

              const newSource = sourceMap
                ? new webpack.sources.SourceMapSource(
                    content,
                    resourceName,
                    sourceMap
                  )
                : new webpack.sources.OriginalSource(content, resourceName);

              finalSource.add(newSource);
              finalSource.add("\n");
              finalSource.add(";\n");
            });

            finalSource.add(new webpack.sources.CachedSource(source));

            return finalSource;
          }
        );
      }
    );
  }
}
