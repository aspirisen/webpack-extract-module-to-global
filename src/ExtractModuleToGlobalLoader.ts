import * as crypto from "crypto";
import type { Compiler } from "webpack";
import { ExtractModuleToGlobal } from "./ExtractModuleToGlobal";

export function ExtractModuleToGlobalLoader(
  this: {
    _compiler: Compiler;
    resource: string;
    getLogger: () => { error: (message: string) => void };
  },
  content: string,
  map: any
) {
  const logger = this.getLogger();

  const extractModuleToGlobal = this._compiler.options.plugins.find(
    (p) => p instanceof ExtractModuleToGlobal
  ) as ExtractModuleToGlobal;

  if (extractModuleToGlobal) {
    extractModuleToGlobal.addSource(this.resource, content, map);
  } else {
    logger.error(`ExtractModuleToGlobal plugin has not been found`);
  }

  const hash = crypto.createHash("md5").update(content).digest("hex");

  return `/** Extracted to global space ${hash} **/`;
}

export default ExtractModuleToGlobalLoader;
