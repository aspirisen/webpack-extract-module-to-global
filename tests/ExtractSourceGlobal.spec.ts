import * as path from "path";
import webpack from "webpack";
import prettier from "prettier";
import { createFsFromVolume, Volume } from "memfs";
import { ExtractModuleToGlobal } from "../src";

describe("ExtractModuleToGlobal", () => {
  it("should extract module", (done) => {
    const compiler = webpack({
      mode: "development",
      context: path.resolve(__dirname, "./spec-app"),
      entry: "./index.ts",
      target: ["web", "es5"],
      output: {
        path: "/",
        filename: "out.js",
      },
      resolve: {
        extensions: [".ts", ".tsx", ".js"],
      },
      module: {
        rules: [
          {
            include: /a.ts/,
            loader: ExtractModuleToGlobal.loader,
          },
          {
            test: /\.tsx?$/,
            loader: "ts-loader",
          },
        ],
      },
      devtool: false,
      plugins: [new ExtractModuleToGlobal()],
    });

    const fs = createFsFromVolume(new Volume());
    compiler.outputFileSystem = fs;

    compiler.run(() => {
      const content = prettier.format(fs.readFileSync("/out.js").toString(), {
        parser: "babel",
      });

      expect(content).toMatchSnapshot();

      const scope = {} as { a: () => string };
      Function(content + "; this.a = a").call(scope);

      expect(scope.a()).toBe("from global space");

      done();
    });
  });
});
