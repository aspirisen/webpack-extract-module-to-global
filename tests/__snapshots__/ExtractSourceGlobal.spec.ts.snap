// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtractModuleToGlobal should extract module 1`] = `
"\\"use strict\\";
function a() {
  return \\"from global space\\";
}

/******/ (function () {
  // webpackBootstrap
  /******/ var __webpack_modules__ = {
    /***/ \\"./b.ts\\":
      /*!**************!*\\\\
  !*** ./b.ts ***!
  \\\\**************/
      /***/ function (__unused_webpack_module, exports) {
        \\"use strict\\";

        Object.defineProperty(exports, \\"__esModule\\", { value: true });
        exports.b = void 0;
        function b() {
          return \\"In module space\\";
        }
        exports.b = b;

        /***/
      },

    /***/ \\"./index.ts\\":
      /*!******************!*\\\\
  !*** ./index.ts ***!
  \\\\******************/
      /***/ function (__unused_webpack_module, exports, __webpack_require__) {
        \\"use strict\\";

        var __createBinding =
          (this && this.__createBinding) ||
          (Object.create
            ? function (o, m, k, k2) {
                if (k2 === undefined) k2 = k;
                Object.defineProperty(o, k2, {
                  enumerable: true,
                  get: function () {
                    return m[k];
                  },
                });
              }
            : function (o, m, k, k2) {
                if (k2 === undefined) k2 = k;
                o[k2] = m[k];
              });
        var __exportStar =
          (this && this.__exportStar) ||
          function (m, exports) {
            for (var p in m)
              if (
                p !== \\"default\\" &&
                !Object.prototype.hasOwnProperty.call(exports, p)
              )
                __createBinding(exports, m, p);
          };
        Object.defineProperty(exports, \\"__esModule\\", { value: true });
        __webpack_require__(/*! ./a */ \\"./a.ts\\");
        __exportStar(__webpack_require__(/*! ./b */ \\"./b.ts\\"), exports);

        /***/
      },

    /***/ \\"./a.ts\\":
      /*!**************!*\\\\
  !*** ./a.ts ***!
  \\\\**************/
      /***/ function () {
        /** Extracted to global space 2d7cedc3557d25f324edf914eac861fe **/
        /***/
      },

    /******/
  };
  /************************************************************************/
  /******/ // The module cache
  /******/ var __webpack_module_cache__ = {};
  /******/
  /******/ // The require function
  /******/ function __webpack_require__(moduleId) {
    /******/ // Check if module is in cache
    /******/ var cachedModule = __webpack_module_cache__[moduleId];
    /******/ if (cachedModule !== undefined) {
      /******/ return cachedModule.exports;
      /******/
    }
    /******/ // Create a new module (and put it into the cache)
    /******/ var module = (__webpack_module_cache__[moduleId] = {
      /******/ // no module.id needed
      /******/ // no module.loaded needed
      /******/ exports: {},
      /******/
    });
    /******/
    /******/ // Execute the module function
    /******/ __webpack_modules__[moduleId].call(
      module.exports,
      module,
      module.exports,
      __webpack_require__
    );
    /******/
    /******/ // Return the exports of the module
    /******/ return module.exports;
    /******/
  }
  /******/
  /************************************************************************/
  /******/
  /******/ // startup
  /******/ // Load entry module and return exports
  /******/ // This entry module is referenced by other modules so it can't be inlined
  /******/ var __webpack_exports__ = __webpack_require__(\\"./index.ts\\");
  /******/
  /******/
})();
"
`;
